File: 代码教程/1、OC基础/Demo_pushVC/Demo_pushVC/BViewController.m
//
//  BViewController.m
//  Demo_pushVC
//
//  Created by gaoguangxiao on 2021/8/11.
//

#import "BViewController.h"
#import "ViewController.h"
@interface BViewController ()


@end

@implementation BViewController

__weak id reference = nil;

- (void)viewWillAppear:(BOOL)animated {
    [super viewWillAppear:animated];
    NSLog(@"B - viewWillAppear - 2");
    NSLog(@"viewWillAppear：%@",reference);
}

- (void)viewDidAppear:(BOOL)animated {
    [super viewDidAppear:animated];
    NSLog(@"B - viewDidAppear - 3");
    NSLog(@"viewDidAppear：%@",reference);
}

- (void)viewDidDisappear:(BOOL)animated {
    [super viewDidDisappear:animated];
    NSLog(@"B - viewDidDisappear -  5");
    
    
}

- (void)viewWillDisappear:(BOOL)animated {
    [super viewWillDisappear:animated];
    NSLog(@"B - viewWillDisappear - 4");
}



- (void)viewDidLoad {
    [super viewDidLoad];
    // Do any additional setup after loading the view.
    NSLog(@"B - viewDidLoad - 1");
    
    NSString *str = [NSString stringWithFormat:@"sunnyxx"];   // str是一个autorelease对象，设置一个weak的引用来观察它
    reference = str;
    
    // viewdidload viewwillappre viewdidaore viewwilldiaapp viewdiddisa deaoolc
}

//- (void)loadView {
//    NSLog(@"B - loadView");
//}

- (IBAction)POPAction:(id)sender {
    
//    [self dismissViewControllerAnimated:YES completion:^{
//
//    }];
    
//    ViewController *bv = [[ViewController alloc]init];
    
//    bv.delegate = self;
    
//    [self.navigationController pushViewController:bv animated:YES];
    
//    [self.navigationController popViewControllerAnimated:YES];
}


- (void)dealloc
{
    NSLog(@"B - dealloc - 6");
    NSLog(@"dealloc：%@",reference);
}
/*
#pragma mark - Navigation

// In a storyboard-based application, you will often want to do a little preparation before navigation
- (void)prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender {
    // Get the new view controller using [segue destinationViewController].
    // Pass the selected object to the new view controller.
}
*/

@end

File: 代码教程/9、性能优化/002-Timer循环引用/002-Timer循环引用/BViewController.swift
//
//  BViewController.swift
//  002-Timer循环引用
//
//  Created by 高广校 on 2023/12/21.
//  处理这种界面销毁

import UIKit

class BViewController: UIViewController {

    var lineTimer :Timer?
    
    override func viewDidLoad() {
        super.viewDidLoad()

        // Do any additional setup after loading the view.
//        initTimer()
        //循环引用的问题是，控制器引用timer，解决方案，建立一个继承自proxy的中间类
//        initScheduleTimer()
//        initProxyTimer()
//        self.initBlockTimer()
        initextensionTimer()
    }
    
    func initTimer() {
        /**
         init(timeInterval...) 创建的timer需要添加runloop中
         */
        lineTimer = Timer(timeInterval: 1.0, target: self, selector: #selector(updateLineTime), userInfo: nil, repeats: true)
        if let lineTimer = lineTimer {
//           由于主线程中的RunLoop是常驻内存同时对Timer的强引用，Timer是局部变量依然会产生循环引用问题
//           Timer继续对控制器强引用
            RunLoop.current.add(lineTimer, forMode: .common)
        }
        
    }
    
    func initScheduleTimer() {
        //已经添加到runloop中，但仍旧有循环引用问题
        let lineTimer = Timer.scheduledTimer(timeInterval: 1.0, target: self, selector: #selector(updateLineTime), userInfo: nil, repeats: true)

    }
    
    func initProxyTimer() {
        let pro = SwiftWeakProxy(self)
        lineTimer = Timer.scheduledTimer(timeInterval: 1.0, target: pro, selector: #selector(updateLineTime), userInfo: nil, repeats: true)
    }
    
    func initBlockTimer() {
    
        lineTimer = Timer.init(timeInterval: 1.0, repeats: true, block: { [weak self]  timer in
            guard let `self` = self else { return }
            self.updateLineTime()
        })
        if let lineTimer = lineTimer {
            RunLoop.current.add(lineTimer, forMode: .common)
        }
    }
    
    func initextensionTimer()  {
        lineTimer = Timer.zk_scheduledTimer(timeInterval: 1.0, repeats: true, with: { [weak self]  in
            guard let `self` = self else { return }
            self.updateLineTime()
        })
    }
    
    @objc func updateLineTime()  {
        print("定时器--调用")
        
        let timerInterval = lineTimer?.timeInterval
        print("lineTimer?.timeInterval:\(timerInterval)")
//        lineTimer?.timeInterval
    }
    
    deinit {
        lineTimer?.invalidate()//
        print("\(self)-deinit")
    }
    
    /*
    // MARK: - Navigation

    // In a storyboard-based application, you will often want to do a little preparation before navigation
    override func prepare(for segue: UIStoryboardSegue, sender: Any?) {
        // Get the new view controller using segue.destination.
        // Pass the selected object to the new view controller.
    }
    */

}
